### PR TITLE
Cli rework in argparse

### PR DESCRIPTION
**Pull request by meng-sun**
All functions have been rewritten with argparsers and tab competition has been added. The changes are in cli_new.py instead of overwriting cli.py.  It's a very long pull request, so please squash the changes! Recursive deletes and confirmation statements have not been added. Documentation is found inside the files.

### CODE REVIEW

**Comment by shwsun** [comment]
This is a huge PR so here are some suggestions that I think can reduce the pain of the reviewers:
* can you guys explain the logic and demo it a little during today's HIL weekly meeting?
* We need documentation for everything, one important thing for open source project is to provide enough documentations plus testing to make sure people can understand the changes and it will not break other running stuff. So please spend some time one doc and unit tests

**Comment by henn** [comment]
To make things more modular, it will be helpful to break apart CommandListener()'s `__init__`, such that each subparser is in its own function.

File: haas/cli_new.py [hunk 1]
```diff
@@ -0,0 +1,1250 @@
...
+        # node parsers
+        node_reg_parser = node_subparsers.add_parser(
+                               'register', parents=[get_name])
+        node_register_subtype = node_reg_parser.add_mutually_exclusive_group(
+                                required=True)
+        node_register_subtype.add_argument('--mock', nargs=3,
+                                           metavar=('host',
+                                                    'user',
+                                                    'password'))
+        node_register_subtype.add_argument('--impi', nargs=3,
```
Comment (naved001) [inline haas/cli_new.py]: typo: should be `ipmi`, not `impi`

**Comment by henn** [comment]
We should support the "help" command too.

Also, a rebase on top of the newest master will get rid of the SQLALCHEMY_TRACK_... errors.

```
$ haas --help
/home/henn/hil-cli/venv/lib/python2.7/site-packages/flask_sqlalchemy/__init__.py:800: UserWarning: SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and will be disabled by default in the future.  Set it to True to suppress this warning.
  warnings.warn('SQLALCHEMY_TRACK_MODIFICATIONS adds significant overhead and will be disabled by default in the future.  Set it to True to suppress this warning.')
usage: haas

positional arguments:
  {serve,serveNetworks,admin,node,network,user,project,switch,headnode,nic,hnic,port}
    serve               starts a http client
```

**Comment by naved001** [comment]
What does `command_listener.py` do? It's not imported or called anywhere. Removing it didn't make any difference either.

**Comment by gfaline** [comment]
Command Listener can be deleted. We initially attempted to make the additions in a separate file for easier review but it did not work out. We used it during to test some things and I guess forgot about it. Sorry

**Comment by naved001** [comment]
@gfaline Thanks for clarifying that!

**Comment by naved001** [comment]
Here are some of my concerns:

## Missing commands:

commands for `node_power_off` and `node_power_cycle` aren't there (or at least I couldn't find those) 

## `haas node connect` doesn't work;

I wanted to put my node under a project. 

So on typing `haas node connect` I am presented with 

```
usage: haas node connect [-h]
                         (--network <network name> <nic name> | --project <project name>)
```

(notice there's no `<obj name>` in the usage here)
and when I follow up with `haas node connect --project MYPROJECTNAME MYNODENAME` it throws an error 

```
AttributeError: 'Namespace' object has no attribute 'name'
```
* Similary, `haas node connect --network MYNETWORKNAME MYNICNAME` throws an AttributeError. 

* However, when I typed `haas project connect --node MYNODENAME MYPROJECTNAME` it worked perfectly! 

* Also, `haas node disconnect` works as expected!

## `haas nic`

```
usage: haas nic [-h] {register,delete,connect,disconnect} ...
haas nic: error: too few arguments
```

I entered `haas nic disconnect` to see `usage:` but it returned `Argument Type Not Accepted`

The user is left to figure out how to use it.

## `haas`

typing `haas` just says 
```
usage: haas
haas: error: too few arguments
``` 
perhaps, we could show the help message here we get when we type `haas --help`

## documentation

As @shwsun said, we really need some sort of documentation. Looking at the code and/or trying to figure out how the commands work is not very efficient. So at the moment I could only test a small portion of it. 

Overall, tab completion is pretty cool and when it works it makes things much faster! Thank you so much for doing this! :)

**Comment by meng-sun** [comment]
Command_listener is actually used in scripts/haas. It loads all the parser structure so that tab completion works without having to import everything required for API access.

**Comment by meng-sun** [comment]
I'll be working on updating the documentation soon! Also, Gwen, feel free to chip in since you know this stuff well :)

**Comment by meng-sun** [comment]
I'm closing this pull request and opening a new pull request for just the user functions as per the new plan!

**Pull request closed without merging**